### PR TITLE
Add official nio and test

### PR DESCRIPTION
**Pull request by jean-philippe-martin**
Add a dependency to the official NIO jar and include tests for gcloud-java-nio.

This is based on the earlier NIO example. Now that NIO support is official we can move this test to our main branch.

CC: @lbergelson @droazen

### CODE REVIEW

**Comment by pgrosu** [comment]
@jean-philippe-martin are you using the proper fork?  There are commits most likely that are not part of your PR.

**Comment by droazen** [comment]
@jean-philippe-martin Can you try squashing the branch into a single commit using a `rebase -i`, and see if the conflict persists?

**Comment by jean-philippe-martin** [comment]
Fixed the merge issue (I had initially rebased off of my local master, which was a little bit behind upstream/master).

**Comment by jean-philippe-martin** [comment]
@droazen how about I first push to a new branch on broadinstitute:, so we can run the Travis tests on this code?

**Comment by pgrosu** [comment]
@jean-philippe-martin Are your jars properly set up?  I'm seeing the Travis build fail the following way:

```
...
Gradle suite > Gradle test > org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSourceUnitTest.testReadFromFileAndHDFS FAILED
    org.apache.hadoop.ipc.RemoteException(java.lang.NoClassDefFoundError): Could not initialize class org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos
    	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol.getDescriptor(ClientNamenodeProtocolProtos.java)
    	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.getDescriptorForType(ClientNamenodeProtocolProtos.java)
    	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:598)
...
```

File: src/main/resources/org/broadinstitute/hellbender/utils/private_file.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+This file is meant to be copied to GCS so that NioTests can make sure they can get to it, even if it's not public.
```
Comment (droazen) [inline src/main/resources/org/broadinstitute/hellbender/utils/private_file.txt]: Should be in `src/test/resources`, I think

**Comment by droazen** [comment]
@jean-philippe-martin Any thoughts on the cause of the failure to load class `org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos` in travis? Do tests pass for you if you run them locally using `gradle test`?

File: src/main/resources/org/broadinstitute/hellbender/utils/private_file.txt (hunk 1 above)
Comment (jean-philippe-martin) [inline src/main/resources/org/broadinstitute/hellbender/utils/private_file.txt]: Done.

**Comment by jean-philippe-martin** [comment]
@droazen I tried the tests locally but I still have to install a bunch of things for them to work. I expect to have time for it later this week.

**Comment by droazen** [comment]
@jean-philippe-martin Ok, ping me once tests are passing and I'll finish the review.

**Comment by jean-philippe-martin** [comment]
Investigating. It looks like the problem has something to do with the hadoop jars using protobuf-java [version 2.x](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-hdfs/2.7.2), but gcloud-java using [version 3.x](https://mvnrepository.com/artifact/com.google.cloud/gcloud-java-core/0.2.5). When we include both dependencies, gradles bumps everything to protobuf 3.x and this seems to break the Hadoop libraries.

I can reproduce the problem on master without even adding a gcs dependency, just by adding a protobuf 3.x dependency.

```
$ git diff build.gradle
+    compile 'com.google.protobuf:protobuf-java:3.0.0-beta-2'
$ ./gradlew clean test --tests '*ReadsSparkSourceUnitTest.testReadFromFileAndHDFS'
Gradle suite > Gradle test > org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSourceUnitTest.testReadFromFileAndHDFS FAILED
    org.apache.hadoop.ipc.RemoteException(java.lang.NoClassDefFoundError): Could not initialize class org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos
```

The problem does not reproduce (at least not for this particular test) if I use an earlier version of the protobuf jar (3.0.0-beta-1). I just tried the whole test suite and it passed.

So one hypothesis is that we may be able to move forward if I can somehow force gcloud-java-core:0.2.5 to use protobufs beta-1 instead of beta-3, and if this change doesn't break the gcloud code.

**Comment by davidbernick** [comment]
Can one of the admins verify this patch?

**Comment by lbergelson** [comment]
verified

**Comment by lbergelson** [comment]
@davidbernick What's the magic statement to make jenkins accept the patch to build? / How can we whitelist people?

**Comment by jean-philippe-martin** [comment]
Guys it's too early to run this on the testing infrastructure, I'm still working on getting it to work locally.

**Comment by jean-philippe-martin** [comment]
This one passes locally, at least for the non-cloud tests.

**Comment by davidbernick** [comment]
@lbergelson you shouldn't see it again but the magic phrases are

"ok to test" to accept this pull request for testing
"test this please" for a one time test run
"add to whitelist" to add the author to the whitelist

**Comment by jean-philippe-martin** [comment]
Trying #2013 to get the cloud regression tests to run.

**Comment by droazen** [comment]
Closing in favor of https://github.com/broadinstitute/gatk/pull/2013

**Pull request closed without merging**